### PR TITLE
Update to cranelift 0.57.0

### LINKED ISSUE #841: Trap handling dereferences a null pointer
Moved from https://github.com/bytecodealliance/wasmtime/issues/829

Running on a Linux machine, using 448faed5ca0e38c3ee5cdbe703b9f89f1e76e8a0 I am consistently observing a segfault caused by dereferencing a null pointer somewhere in libgcc while creating a backtrace. Wasmtime is used through `wasmtime-api`.

```
* thread #1, name = 'sc_executor-f4f', stop reason = signal SIGSEGV: invalid address (fault address: 0x8)
  * frame #0: 0x00007ffff72185df libgcc_s.so.1`___lldb_unnamed_symbol41$$libgcc_s.so.1 + 1055
    frame #1: 0x00007ffff721a138 libgcc_s.so.1`_Unwind_Backtrace + 104
    frame #2: 0x0000555555a68137 sc_executor-f4fe4729a78e1cc7`backtrace::backtrace::trace::h1adcaa00054f7e2c + 87
    frame #3: 0x0000555555a67a56 sc_executor-f4fe4729a78e1cc7`backtrace::capture::Backtrace::new_unresolved::hcc88b2186746cdd0 + 86
    frame #4: 0x0000555555710c9f sc_executor-f4fe4729a78e1cc7`RecordTrap + 239
    frame #5: 0x00005555557128ff sc_executor-f4fe4729a78e1cc7`WasmTrapHandler(int, siginfo_t*, void*) + 431
    frame #6: 0x00007ffff7433890 libpthread.so.0`___lldb_unnamed_symbol1$$libpthread.so.0 + 1
    frame #7: 0x00007ffff7fb3d84
    frame #8: 0x00007ffff7fb365a
    frame #9: 0x00007ffff7fb06a5
    frame #10: 0x00007ffff7fbe2f2
    frame #11: 0x00007ffff67f9031
    frame #12: 0x0000555555712b40 sc_executor-f4fe4729a78e1cc7`WasmtimeCallTrampoline + 128
    frame #13: 0x0000555555711517 sc_executor-f4fe4729a78e1cc7`wasmtime_call_trampoline + 23
...
```

the code around there looks like this

```
    0x7ffff72185d8 <+1048>: movq   0x98(%r13), %rax
->  0x7ffff72185df <+1055>: cmpb   $0x48, (%rax)
    0x7ffff72185e2 <+1058>: jne    0x7ffff7218580            ; <+960>
    0x7ffff72185e4 <+1060>: movabsq $0x50f0000000fc0c7, %rdx  ; imm = 0x50F0000000FC0C7 
    0x7ffff72185ee <+1070>: cmpq   %rdx, 0x1(%rax)
    0x7ffff72185f2 <+1074>: jne    0x7ffff7218580            ; <+960>
```

The original PR https://github.com/paritytech/substrate/pull/4686

### PR DESCRIPTION
**Pull request by alexcrichton**
Still needs a crates.io publish, but figure we can throw CI at this
change in the meantime!

Closes #841

### CODE REVIEW

File: crates/environ/src/func_environ.rs [hunk 1]
```diff
@@ -799,4 +799,69 @@ impl<'module_environment> cranelift_wasm::FuncEnvironment for FuncEnvironment<'m
...
+    ) -> WasmResult<ir::Value> {
+        Err(WasmError::Unsupported("reference types".to_string()))
+    }
+
+    fn translate_custom_global_get(
+        &mut self,
+        _pos: FuncCursor,
+        _global_index: GlobalIndex,
+    ) -> WasmResult<ir::Value> {
+        Err(WasmError::Unsupported("reference types".to_string()))
```
Comment (fitzgen) [inline crates/environ/src/func_environ.rs]:
```suggestion
        unreachable!("we don't create any custom globals")
```

File: crates/environ/src/func_environ.rs [hunk 2]
```diff
@@ -799,4 +799,69 @@ impl<'module_environment> cranelift_wasm::FuncEnvironment for FuncEnvironment<'m
...
+    }
+
+    fn translate_custom_global_set(
+        &mut self,
+        _pos: FuncCursor,
+        _global_index: GlobalIndex,
+        _val: ir::Value,
+    ) -> WasmResult<()> {
+        Err(WasmError::Unsupported("reference types".to_string()))
```
Comment (fitzgen) [inline crates/environ/src/func_environ.rs]:
```suggestion
        unreachable!("we don't create any custom globals")
```

**Comment by alexcrichton** [comment]
Oh oops my bad, didn't mean to scoop a PR from you :(

**Comment by fitzgen** [comment]
not an issue ;)

**Comment by fitzgen** [comment]
Ignoring that the `frame_size` field apparently doesn't exist anymore, this snippet looks very cursed... why would we ever be assigning a random value to a frame size??

```
error[E0609]: no field `frame_size` on type `cranelift_codegen::ir::stackslot::StackSlots`
   --> crates/environ/src/cache/tests.rs:320:19
    |
320 |             slots.frame_size = rng.gen();
    |                   ^^^^^^^^^^ unknown field
    |
    = note: available fields are: `layout_info`
```

**Comment by fitzgen** [comment]
Anyways, deleting that line fixes the compilation errors and the test passes.

**Comment by fitzgen** [comment]
Does it make sense to land this with the git deps until someone with crates.io ownership publishes cranelift 0.57.0? Doing so would unblock #920 to land as well.

**Comment by fitzgen** [comment]
We just published 0.58.0 and I'm folding that into #926

**Pull request closed without merging**